helion/language/loops.py

```diff
@@ -449,6 +451,96 @@ def _(state: CodegenState) -> ast.AST:
+    if isinstance(value, (list, tuple)):
+        return list(value)
+    return [value]
+
+
+def _normalize_begin_end_ref(
+    begin_or_end: int | torch.Tensor | list[int | torch.Tensor],
+    end_or_none: int | torch.Tensor | list[int | torch.Tensor] | None = None,
+) -> tuple[
+    int | torch.Tensor | list[int | torch.Tensor],
+    int | torch.Tensor | list[int | torch.Tensor],
+]:
+    if end_or_none is not None:
+        # Two positional args: begin_or_end is begin, end_or_none is end
+        return begin_or_end, end_or_none
+    # One positional arg: begin_or_end is end, begin defaults to 0
+    end = begin_or_end
+    if isinstance(end, (list, tuple)):
+        begin = cast("int | torch.Tensor | list[int | torch.Tensor]", [0] * len(end))
+    else:
+        begin = 0
+    return begin, end
+
+
+@_decorators.ref(tile)
+def _(
+    begin_or_end: int | torch.Tensor | list[int | torch.Tensor],
+    end_or_none: int | torch.Tensor | list[int | torch.Tensor] | None = None,
+    block_size: int | torch.Tensor | list[int | torch.Tensor] | None = None,
+) -> Iterator[RefTile | tuple[RefTile, ...]]:
+    # Step 1: Normalize begin and end values
+    begin, end = _normalize_begin_end_ref(begin_or_end, end_or_none)
+
+    # Step 2: Convert to lists and then to ints
+    begin_list = _normalize_to_list(begin)
+    end_list = _normalize_to_list(end)
+    begin_ints = [_to_int(b) for b in begin_list]
+    end_ints = [_to_int(e) for e in end_list]
+
+    # Step 3: Determine block sizes - always return full dimension size, ignoring block_size parameter
+    block_size_list = []
+    for b, e in zip(begin_ints, end_ints, strict=True):
+        assert b is not None and e is not None
+        block_size_list.append(e - b)
+
+    # Step 4: Determine return type
+    # Return single tiles if input was not a list
+    return_single = not isinstance(begin, list) and not isinstance(end, list)
+
+    # Step 5: Generate tiles
+    if len(begin_ints) == 1 and return_single:
+        # Single dimension case
+        assert begin_ints[0] is not None and end_ints[0] is not None
+        begin = begin_ints[0]
+        end = end_ints[0]
+        if begin != end:
+            # Only yield one tile covering the full range
+            yield RefTile(begin, end, end - begin)
+    else:
+        # Multi-dimensional case
+        assert all(b is not None for b in begin_ints)
+        assert all(e is not None for e in end_ints)
+
+        # Generate a single tile per dimension covering the full range
+        ranges = []
+        for b, e, bs in zip(begin_ints, end_ints, block_size_list, strict=True):
+            assert b is not None and e is not None and bs is not None
+            if b == e:
+                # Empty range for this dimension, skip
+                continue
+            # Only one tile covering full dimension
+            dim_tiles = [RefTile(b, e, e - b)]
+            ranges.append(dim_tiles)
+
+        yield from itertools.product(*ranges)
```

Comment:
I don't think this product is needed anymore with a single tile per dimension.  Can we unify the code between the two cases?